networkx/algorithms/community/asyn_lpa.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+#    Copyright (C) 2015
+#    All rights reserved.
+#    BSD license.
+
+import random
+import networkx as nx
+
+
+def asyn_lpa_communities(G, weight=None):
+    """
+    Asynchronous label propagation algorithm for community detection described
+    on [1]_. The algorithm is non-deterministic and the found communities may
+    vary on different executions.
+
+    The algorithm follows these steps:
+
+        1. Initialize the labels at all nodes in the network giving them unique
+        identifiers.
+        2. Set t=1.
+        3. Arrange the nodes in the network in a random order and set it to X.
+        4. For each x in X chosen in that specific order, get the label
+        occurring with the highest frecuency among neighbors. Ties are broken
+        uniformly randomly.
+        5. If every node has a label that the maximun number of their neighbors
+        have, then stop the algorithm. Else set t = t+1 and got to (3).
+
+    Modified to be able to take into account the edge weights when assigning
+    the labels to one node.
+
+    Parameters
+    ----------
+    G : graph
+        Network to be processed
+
+    weight : string
+        If the graph is weighted, the edge parameter representing the weight.
+        If left to None edge weights are not taken into account.
+
+    Returns
+    -------
+    communities : list
+        List of communities (sets)
+
+    References
+    ----------
+    .. [1] Raghavan, Usha Nandini, Réka Albert, and Soundar Kumara. "Near
+           linear time algorithm to detect community structures in large-scale
+           networks." Physical Review E 76.3 (2007): 036106.
+    """
+
+    labels = {n: i for i, n in enumerate(G)}
+    cont = True
+    while cont:
+        cont = False
+        nodes = nx.nodes(G)
+        random.shuffle(nodes)
+        # Calculate the label for each node
+        for node in nodes:
+            n_neighbors = nx.neighbors(G, node)
+
+            if len(n_neighbors) < 1:
+                continue
+
+            # Get label frecuencies. Depending on the order they are processed
+            # in some nodes with be in t and others in t-1, making the
+            # algorithm asynchronous
+            label_freq = {}
```

Comment:
Use `label_freq = defaultdict(lambda: 0)` so that you can just use `label_freq += G.edge[neighbor][node][weight] if weight else 1` below instead of the nested if/else blocks.